### PR TITLE
Add StableAdamW Optimizer

### LINKED ISSUE #36564: Add support for StableAdamW optimizer in Trainer
### Feature request

StableAdamW is an optimizer first introduced in [Stable and low-precision training for large-scale vision-language models](https://arxiv.org/pdf/2304.13013), an AdamW and AdaFactor hybrid optimizer, leading to more stable training. Most notably, however, it has been used in the [modernBERT paper](https://arxiv.org/pdf/2412.13663):
> StableAdamW’s learning rate clipping outperformed standard gradient clipping on downstream tasks and led to more stable training

It would be great is this is available as an optimizer in `Trainer`!

### Motivation

More models in the future may use StableAdamW because of its success in training modernBERT, and having it as an option in `Trainer` (as `optim` in `TrainingArguments`) would be convenient.

### Your contribution

I'm interested to contribute! The modernBERT paper uses the implementation from [optimi](https://github.com/warner-benjamin/optimi), which can be added as an import. I'd love to submit a PR.

### PR DESCRIPTION
**Pull request by capemox**
# What does this PR do?
This PR adds the `StableAdamW` optimizer from the [`optimi`](https://optimi.benjaminwarner.dev/) package. In this PR:
- StableAdamW is added to `OptimizerNames`
- The optimizer can be initialized through `TrainingArguments`, gets imported and used in `Trainer`
- Test cases are written to verify the proper behaviour of the model under `TrainerIntegrationTest`

Fixes #36564 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?
@muellerzr @SunMarc should be the right people to review this!

### CODE REVIEW

**Comment by capemox** [comment]
The force push was to remove a commit I accidentally included in the PR, sorry!

**Comment by Rocketknight1** [comment]
cc @muellerzr @SunMarc for review (but feel free to push the review off onto someone else if you think they're better suited!)

**Review by SunMarc — APPROVED** [review]
Thanks ! Can you update a bit the docs ?

**Comment by capemox** [comment]
@SunMarc, I've updated the docs. Thanks for the approval!

**Review by shethaadit — APPROVED** [review]

**Comment by walter-hernandez** [comment]
Some of the tests are failing, which are what is remaining to finish the merge of this PR into main. What is the process to solve for the pending tests?

**Comment by walter-hernandez** [comment]
@SunMarc For the tests, it looks like the workflow `run_tests` is failing here: `tests/models/opt/test_modeling_opt.py::OPTModelTest::test_pipeline_text_classification` which generates `Failed: Timeout >120.0s` for `hf_hub_download()` .  Maybe increase the timeout to give more time to download? This is the remaining test to pass to be able to merge into main. Thank you @SunMarc !

**Comment by SunMarc** [comment]
I've created this PR to fix the github actions that is broken here https://github.com/huggingface/transformers/pull/39446. So i'm closing this PR !
